### PR TITLE
Added support for setting owned resources

### LINKED ISSUE #18: Add support for uninstalling Gatekeeper
When the Gatekeeper CR named `gatekeeper` is deleted, all of the Gatekeeper related resources should be subsequently deleted. This is essentially the reverse of the installation path.

### PR DESCRIPTION
**Pull request by sabre1041**
Added support for setting owned resources to resolve #18

### CODE REVIEW

**Review by JAORMX — APPROVED** [review]
LGTM

File: controllers/gatekeeper_controller.go
Comment (font) [inline controllers/gatekeeper_controller.go:160]: We should check the error returned here and report any issues.

Comment (font) [inline controllers/gatekeeper_controller.go:160]: I don't believe this will work for deleting Gatekeeper's cluster-scoped resources e.g. CRDs, validating webhook configuration, cluster role and cluster role binding, because the gatekeeper CR is namespace-scoped and cannot be an owner for cluster-scoped resources.

**Review by font — COMMENTED** [review]
Thanks @sabre1041! I made a couple comments.

File: controllers/gatekeeper_controller.go
Comment (JAORMX) [reply controllers/gatekeeper_controller.go:160]: Right, that could be achieved via finalizers. @sabre1041 let me know if you want to add finalizer support to delete the cluster-scoped resources. Else, I could start that work after this PR merges once some details are fixed.

Comment (sabre1041) [reply controllers/gatekeeper_controller.go]: @font @JAORMX Added finalizer and addressed remaining issues

**Review by JAORMX — APPROVED** [review]

File: controllers/gatekeeper_controller.go
Comment (font) [inline controllers/gatekeeper_controller.go:293]: We can simplify this by using the apimachinery sets package https://godoc.org/k8s.io/apimachinery/pkg/util/sets#String. We just create a new string set `NewString()` containing the finalizers from `GetFinalizers()` and then invoke the `Has()` API method.

File: controllers/gatekeeper_controller.go
Comment (font) [inline controllers/gatekeeper_controller.go:287]: For consistency, I think we should wrap this error with this message and return the wrapped error.

File: controllers/gatekeeper_controller.go
Comment (font) [inline controllers/gatekeeper_controller.go:212]: nit: s/reference/reference for/

File: controllers/gatekeeper_controller.go
Comment (font) [inline controllers/gatekeeper_controller.go:270]: nit: remove extra line here and after next closing brace.

File: controllers/gatekeeper_controller.go
Comment (font) [inline controllers/gatekeeper_controller.go:157]: Is this needed?

**Review by font — COMMENTED** [review]
Thank you @sabre1041! :+1: I'm looking forward to having uninstalls working. I made a few comments, but I'm happy to merge this as-is, pending my question about the event predicates. We can address the comments in a follow-up PR or in this PR. Let me know how you'd like to proceed.

File: controllers/gatekeeper_controller.go
Comment (sabre1041) [reply controllers/gatekeeper_controller.go:157]: Yes. This prevents reconciling for certain additional unnecessary reconcile loops for certain types of updates

**Comment by sabre1041** [comment]
@font updated based on comments

File: controllers/gatekeeper_controller.go
Comment (font) [reply controllers/gatekeeper_controller.go:158]: After looking more into this, it's not required, but helps reduce unnecessary noise.

**Review by font — APPROVED** [review]

**Pull request merged**
